File: week3/p1.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* insertAtEnd(struct Node* head, int x) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = x;
    newNode->next = NULL;

    if (head == NULL) {
        return newNode; // if list is empty
    }

    struct Node* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->next = newNode;

    return head;
}

// Helper to print linked list
void printList(struct Node* head) {
    while (head != NULL) {
        printf("%d", head->data);
        if (head->next) printf("->");
        head = head->next;
    }
    printf("\n");
}

int main() {
    int n, x;
    scanf("%d", &n);
    struct Node* head = NULL;
    struct Node* tail = NULL;

    for (int i = 0; i < n; i++) {
        int val;
        scanf("%d", &val);
        struct Node* node = (struct Node*)malloc(sizeof(struct Node));
        node->data = val;
        node->next = NULL;
        if (head == NULL) {
            head = tail = node;
        } else {
            tail->next = node;
            tail = node;
        }
    }

    scanf("%d", &x);
    head = insertAtEnd(head, x);
    printList(head);
    return 0;
}

File: week3/p1.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def insertAtEnd(head, x):
    new_node = Node(x)
    if head is None:
        return new_node  # empty list case
    temp = head
    while temp.next:
        temp = temp.next
    temp.next = new_node
    return head

# Helper to print linked list
def printList(head):
    temp = head
    while temp:
        print(temp.data, end="->" if temp.next else "")
        temp = temp.next
    print()

# Example usage
n = int(input())
values = list(map(int, input().split()))
x = int(input())

head = None
tail = None
for val in values:
    node = Node(val)
    if head is None:
        head = tail = node
    else:
        tail.next = node
        tail = node

head = insertAtEnd(head, x)
printList(head)
